### DISCUSSION TITLE
The vision and teaching approach in the app

### DISCUSSION THREAD

**Discussion by NathanLovato** [GDQuest/learn-gdscript#85] [General]
We can discuss the general vision and possible other app projects or improvements in this thread.

This is a continuation of the discussion started in #64.

Note that we'll keep iterating on the current design and content for the coming months. We have to deliver a complete course to our backers quickly and have a fixed budget.

The current design also matches our promise to the Kickstarter backers and community: making an app like FreeCodeCamp or Codecademy.

## The premise and feedback we got

The discussion in #64 started with the suggestion to make a much more gamified app, more in the likes of [Duolingo](https://www.duolingo.com/). An app focused more on coding and bite-sized practices.

If I got it right, the idea is also to bring more _fun_ into the experience (our app is currently more on the serious side).

The premise is that such a gamified app would appeal to people who don't particularly want to become programmers but just want to get into game creation.

In contrast, our app is more for people who want to become developers (more on that below).

One of the main arguments is that people who want to become developers already have documentation, tutorials, and other resources. So they don't need an app as much as this other audience.

_I hope I'm not misrepresenting any argument, otherwise, please let me know, and I'll edit the above. That's what I understood as the key points of the feedback._

I understand there are different approaches to making educational apps and different possible visions.

I think there is room for different kinds of apps revolving around Godot for different target audiences. We already have ideas for other possible apps in the future.

So although we'll keep improving this app and continue on the same path for the reasons above, the discussion on improvements, changes, and possibly other projects is open if you'd like to discuss that.

## The problem, solution, and vision

Here are the foundations everyone needs to understand what we're trying to achieve and what we're doing.

**In short**

Our target audience is explicitly people who don't know how to code but want to *become programmers*. We intend to improve their learning experience with Godot compared to existing resources for the engine and help onboard more people.

**The problem**

The starting point for any project like this is to look at the _painful_ problems the target audience has. 

We have many persons worldwide who want to become developers by getting straight into game development.

They might want to switch careers but be unable to afford university. They might be from a country where they don't have access to a good school or a good learning path. But they want to create and ship games and software.

Many want to use Godot because it's free and open-source (and they're a bit excited by it).

Note that we're not talking about people who want to dabble a bit into code or have a bit of fun.

The thing is, our target audience is lacking foundations and a clear path to learning those foundations. There's no comprehensive course we'd recommend for GDScript.

Instead, we (the Godot community) often recommended people to start with Python in the past. Nobody had the time or budget to really cover GDScript and basic programming concepts in-depth for beginners.

**The solution**

We could've solved that with a plain text-based course to teach GDScript. But people struggle to practice when learning online, even if you give them assignments.

Also, jumping straight into the Godot editor requires learning instantly about the interface, making scenes, nodes, and what-not.

This adds quite a bit of cognitive load on top of the basic programming concepts you need to understand to become independent at coding. Functions, variables, loops, objects, learning to think like a programmer, breaking down problems into smaller ones, debugging, etc.

That's why we chose to make an app. It's still a detailed course where we can cover all the concepts someone who wants to become a developer needs.

But it's also no setup, we can add more interactivity than a plain course, and push users to complete practices.

As you can see above, the intention is to bring substantial improvements to what's available out there for people who are already pretty serious about learning to code.

We think that as a positive side-effect, we'll help promote Godot to many people and professionals who passed due to the lack of a good introductory resource.

You can think of this app as being in the line of FreeCodeCamp and the likes, but for game developers.

## Disclaimer

The content and level of interactivity in the app's first beta release don't fully reflect what we'd like it to be like by the 1.0 release. We're well aware there's a lot to refine and room for many improvements.

We want more interactivity, a more refined UX, but all that takes time and a lot of feedback from people in the target audience.

## Closing words

Hopefully, this gives a better idea of why we're doing the app the way we are. That doesn't mean we won't consider any gamification or making things more playful or more exciting, of course. 

As long as it fits our target and gets students where we want to take them: to the point where they can get into Godot and start creating toys and small games, and look up things they don't know efficiently (using a code reference, etc.).

**Comment by golddotasksquestions** [comment]
Just to clarify, I was not talking about or even thinking of Code Combat. I think these coding games are about as far from duolingo as duolingo is from your app.

**Reply by NathanLovato** [reply]
What would you envision precisely? Writing small code snippets to solve a problem like Duolingo is about writing short sentence translations? Note that I removed code combat from the first post to better represent your idea.

**Reply by golddotasksquestions** [reply]
I opened https://github.com/GDQuest/learn-gdscript/issues/64 to explain what I personally expected vs what I experienced. 
I did not open the 64 issue to say you should throw everything out and make a GDScript-Duolingo app instead, as this would mean making everything anew from scratch basically and would not be constructive. This is just my personal first impression feedback of what I expected and nothing more. 

However regarding the existing app, I still think generally aiming for the least amount of text possible and more interaction would mean improvement ([redblobgames articles](https://www.redblobgames.com/pathfinding/a-star/introduction.html) are a good example for interaciton). I also think only using existing built-in GDScript in the first examples would be important and would not require much change to the course. I also think you should teach variables before functions. Personally I would also touch on update loop and script parsing first, as these two concepts explain why anything you would do with variables and function works in the first place.

**Reply by NathanLovato** [reply]
I didn't take it as if you suggested throwing everything away. More like making the lessons much shorter and leaning more towards lots of examples and small, varied practices (like coding drills).

More interactivity is generally a goal for us. We're also aiming for as little text as possible, although that's always tricky. Students tend to raise lots of questions when testing, so you add answers to those in the text, and it grows. I think we can find a good balance by splitting up longer lessons into smaller bits with more practices in-between.
